buildSrc/subprojects/parser/src/main/java/org/apache/calcite/buildtools/parser/DialectGenerate.java

```diff
@@ -120,10 +124,152 @@ public void unparseReservedKeywords(ExtractedData extractedData) {
       tokens.add(tokenBuilder.toString());
     }
     stringBuilder.append(String.join("| ", tokens));
-    stringBuilder.append("}\n");
+    stringBuilder.append("}");
     extractedData.tokenAssignments.add(stringBuilder.toString());
   }
 
+  /**
+   * Partitions {@code extractedData.nonReservedKeywords} amongst multiple
+   * functions and creates one function that calls all of them of the form:
+   *
+   * // Auto generated.
+   * void NonReservedKeyword():
+   * {
+   * }
+   * {
+   *     (
+   *         NonReservedKeyword0of<numPartitions>()
+   *     |   NonReservedKeyword1of<numPartitions>()
+   *     ...
+   *     )
+   *     { return unquotedIdentifier(); }
+   * }
+   *
+   * If {@code extractedData.nonReservedKeywords} is empty, the function only contains
+   * the return statement.
+   *
+   * Note: Indentation is added here just for clarity, actual function doesn't
+   * have indentation as it is difficult to format using String.join().
+   *
+   * @param extractedData The object which keeps state of all of the extracted
+   *                      data
+   */
+  public void unparseNonReservedKeywords(ExtractedData extractedData) {
+    final ArrayList<Set<Keyword>> partitions = getPartitions(extractedData.nonReservedKeywords);
+    final int numPartitions = partitions.size();
+    final String declarationTemplate = "void NonReservedKeyword%dof"
+      + numPartitions + "():";
+    final StringBuilder bodyBuilder = new StringBuilder();
+    final List<String> functionCalls = new LinkedList<String>();
+    if (!partitions.isEmpty()) {
+      bodyBuilder.append("(\n");
+    }
+    for (int i = 0; i < numPartitions; i++) {
+      String declaration = String.format(declarationTemplate, i, numPartitions);
+      String functionName = getFunctionName(declaration);
+      extractedData.functions.put(functionName,
+          getPartitionFunction(declaration, partitions.get(i)));
+      functionCalls.add(functionName + "()\n");
+    }
+    bodyBuilder.append(String.join("| ", functionCalls));
+    if (!partitions.isEmpty()) {
+      bodyBuilder.append(")\n");
+    }
+    bodyBuilder.append("{ return unquotedIdentifier(); }\n");
+    String declaration = "void NonReservedKeyword():";
+    String function = buildAutoGeneratedFunction(declaration,
+        bodyBuilder.toString());
+    extractedData.functions.put("NonReservedKeyword", function);
+  }
+
+  /**
+   * Creates a string of the form:
+   *
+   * // Auto generated.
+   * <declaration>
+   * {
+   * }
+   * {
+   *         <KEYWORD1>
+   *     |   <KEYWORD2>
+   *     ...
+   * }
+   *
+   * Note: Indentation is added here just for clarity, actual function doesn't
+   * have indentation as it is difficult to format using String.join().
+   *
+   * @param declaration The function declaration
+   * @param keywords The keywords to be in the function body
+   *
+   * @return The formatted string
+   */
+  public String getPartitionFunction(String declaration, Set<Keyword> keywords) {
```

Comment:
nit: Should we mark "keywords" as final?

Reply:
Good idea.